src/sessions.ts

```diff
@@ -523,7 +523,11 @@ export function maybeClearPinnedConnection(
     const servers = Array.from(topology.s.servers.values());
     const loadBalancer = servers[0];
 
-    if (options?.error == null || options?.force) {
+    if (
+      options?.error == null ||
+      options?.error?.name === 'MongoExpiredSessionError' ||
+      options?.force
+    ) {
```

Comment:
Without this change, endSession is sensitive to the exact order of endSession calls that were made by our cursor before this refactor.